File: k12ms26/graph/최단경로.py
#BOJ 1753 최단경로
import heapq
import sys
INF = sys.maxsize

V, E = map(int, sys.stdin.readline().split())
K = int(sys.stdin.readline())
graph = [[] for i in range(V+1)]
# visited = [False] * (V+1)
distance = [INF] * (V+1)

for i in range(E) :
    u, v, w = map(int, sys.stdin.readline().split())
    graph[u].append([v,w])

def dijkstra(start) :
    q = []
    heapq.heappush(q, (0, start))
    distance[start] = 0
    while q:
        print(q)
        dist, now = heapq.heappop(q)
        if distance[now] < dist :
            continue
        for i in graph[now] :
            cost = dist + i[1]
            if cost < distance[i[0]] :
                distance[i[0]] = cost
                heapq.heappush(q, (cost, i[0]))

# def get_smallest_node() :
#     min_value = INF
#     index = 0
#     for i in range(1, V+1) :
#         if distance[i] < min_value and not visited[i] :
#             min_value = distance[i]
#             index = i
#     return index

# def dijkstra(start) :
#     distance[start] = 0
#     visited[start] = True
#     for i in graph[start] :
#         distance[i[0]] = i[1]
#     for i in range(V-1) :
#         now = get_smallest_node()
#         visited[now] = True
#         for j in graph[now] :
#             cost = distance[now] + j[1]
#             if cost < distance[j[0]] :
#                 distance[j[0]] = cost

dijkstra(K)

for i in range(1, V+1) :
    if distance[i] == INF :
        print("INF")
    else :
        print(distance[i])

File: k12ms26/graph/최단경로.js
//BOJ 1753 최단경로
const input = [];
require("readline")
.createInterface(process.stdin, process.stdout)
.on("line", (line) => {
  input.push(line);
})
.on("close", () => {
  solution();
  process.exit();
});

class MinHeap {
    constructor() {
        this.heap = [];
    }

    push(value) {
        this.heap.push(value);
        let curIdx = this.heap.length-1;
        let parIdx = Math.floor(curIdx/2);

        while(parIdx !== 0 && this.heap[parIdx][0] > value[0]) {
            const temp = this.heap[parIdx];
            this.heap[parIdx] = value;
            this.heap[curIdx] = temp;

            curIdx = parIdx;
            parIdx = Math.floor(curIdx/2);
        }
    }
}

const solution = () => {
    const [V, E] = input[0].split(' ').map(Number);
    const K = +input[1];
    let graph = Array.from({length: V+1}, ()=>[]);
    let distance = new Array(V+1).fill(Infinity);

    for(let i=0;i<E;i++) {
        const [u, v, w] = input[i+2].split(' ').map(Number);
        graph[u].push([v, w]);
    }

    const dijkstra = (start) => {
        const heap = new MinHeap();
        heap.push([0, start]);
        distance[start] = 0;

        while(heap.heap.length) {
            let [dist, now] = heap.heap.shift();
            if(distance[now] < dist) {
                continue;
            }
            for(let i=0;i<graph[now].length;i++) {
                let cur = graph[now][i];
                let cost = dist + cur[1];
                if(cost < distance[cur[0]]) {
                    distance[cur[0]] = cost;
                    heap.push([cost, cur[0]]);
                }
            }
        }
    }
    
    dijkstra(K);

    for(let i=1;i<=V;i++) {
        if(distance[i] === Infinity) console.log('INF');
        else console.log(distance[i]);
    }
};